src/dict.c

```diff
@@ -207,34 +206,46 @@ static inline void *decodeMaskedPtr(const dictEntry *de) {
     return (void *)((uintptr_t)(void *)de & ~ENTRY_PTR_MASK);
 }
 
+static inline dictEntry *createEntryNormal(void *key, dictEntry *next) {
+    _dictEntryNormal *entry = zmalloc(sizeof(_dictEntryNormal));
+    assert(entryIsNormal(entry)); /* Check alignment of allocation */
+    entry->key = key;
+    entry->next = next;
+    return encodeMaskedPtr(entry, ENTRY_PTR_NORMAL);
+}
+
 /* Creates an entry without a value field. */
 static inline dictEntry *createEntryNoValue(void *key, dictEntry *next) {
-    dictEntryNoValue *entry = zmalloc(sizeof(*entry));
+    _dictEntryNoValue *entry = zmalloc(sizeof(*entry));
     entry->key = key;
     entry->next = next;
     return encodeMaskedPtr(entry, ENTRY_PTR_NO_VALUE);
 }
 
 static inline dictEntry *createEmbeddedEntry(void *key, dictEntry *next, dictType *dt) {
     size_t key_len = dt->embedKey(NULL, 0, key, NULL);
-    embeddedDictEntry *entry = zmalloc(compactSizeEmbeddedDictEntry() + key_len);
+    _dictEntryEmbedded *entry = zmalloc(compactSizeEmbeddedDictEntry() + key_len);
     dt->embedKey(entry->key_buf, key_len, key, &entry->key_header_size);
     entry->next = next;
     return encodeMaskedPtr(entry, ENTRY_PTR_EMBEDDED);
 }
 
 static inline void *getEmbeddedKey(const dictEntry *de) {
-    embeddedDictEntry *entry = (embeddedDictEntry *)decodeMaskedPtr(de);
+    _dictEntryEmbedded *entry = (_dictEntryEmbedded *)decodeMaskedPtr(de);
     return &entry->key_buf[entry->key_header_size];
 }
 
 /* Decodes the pointer to an entry without value, when you know it is an entry
  * without value. Hint: Use entryIsNoValue to check. */
-static inline dictEntryNoValue *decodeEntryNoValue(const dictEntry *de) {
+static inline _dictEntryNoValue *decodeEntryNoValue(const dictEntry *de) {
     return decodeMaskedPtr(de);
 }
 
-static inline embeddedDictEntry *decodeEmbeddedEntry(const dictEntry *de) {
+static inline _dictEntryEmbedded *decodeEntryEmbedded(const dictEntry *de) {
+    return decodeMaskedPtr(de);
+}
+
+static inline _dictEntryNormal *decodeEntryNormal(const dictEntry *de) {
     return decodeMaskedPtr(de);
 }
```

Comment:
There is some inconsistent assertion usage. It seems like this would be the best place to add
```suggestion
static inline _dictEntryNormal *decodeEntryNormal(const dictEntry *de) {
    assert(entryIsNormal(de));
    return decodeMaskedPtr(de);
}
```

But I think we wanted to avoid this for performance reasons. Do we want to add the compile time version now (not necessarily in this PR, can be a followup)?

Reply:
I think it's being done in all the place to decide if this method should be invoked. Not sure how we want to structure it. But redoing the validation doesn't help.

Reply:
My specific suggestion is because there is an assert before calling decodeEntryNormal() on line 1008, so I think either we omit it everywhere or add the assert everywhere.

Reply:
Will give all the asserts a once-over